optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/bi/DroolsBiGroupBy.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.stream.drools.bi;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiFunction;
+
+import org.drools.core.common.InternalFactHandle;
+import org.optaplanner.core.api.score.stream.bi.BiConstraintCollector;
+import org.optaplanner.core.impl.score.stream.drools.common.BiTuple;
+
+final class DroolsBiGroupBy<A, B, ResultContainer, NewA, NewB> implements Serializable {
+
+    private static final long serialVersionUID = 510l;
+    private final Map<Long, Runnable> undoMap = new HashMap<>(0);
+    private final BiFunction<A, B, NewA> groupKeyMapping;
+    private final BiConstraintCollector<A, B, ResultContainer, NewB> collector;
+    private DroolsBiGroupByAccumulator<A, B, ResultContainer, NewA, NewB> acc;
+
+    public DroolsBiGroupBy(BiFunction<A, B, NewA> groupKeyMapping,
+            BiConstraintCollector<A, B, ResultContainer, NewB> collector) {
+        this.groupKeyMapping = groupKeyMapping;
+        this.collector = collector;
+    }
+
+    public void init() {
+        acc = new DroolsBiGroupByAccumulator<>(groupKeyMapping, collector);
+        undoMap.clear();
+    }
+
+    public void accumulate(InternalFactHandle handle, A a, B b) {
+        Runnable undo = acc.accumulate(a, b);
+        Runnable previousUndo = this.undoMap.put(handle.getId(), undo);
```

Comment:
nitpick code consistently: when optaplanner code captures the return value of map.put() or list.add(), it's typically called old*. => Runnable oldUndo.